doc/amoc_coordinator.md

```diff
@@ -0,0 +1,170 @@
+
+## Description
+
+This module allows to synchronize the users, and act on groups of them.
+
+The coordinator reacts to new users showing up in a system, according to the `Coordination Plan`. 
+The `Coordination Plan` consists of `Coordination Items`, and each of them is defined as one of the following: `{NumberOfUsers, CoordinationActions}`.
+ - When the `Number Of Users` is set to `all`, then only `Coordination Actions` with the arities `/1, /2` are handled. 
+ The `Coordination Items` with `all` are triggered by the `timeout` event. 
+ - When the `Number Of Users` is set to a positive integer, all `Coordination Actions` with arities `/1, /2` and `/3` are handled.
+
+The timeout timer is reset by calling the `add` function.
+A new batch size is set in the `NumberOfUsers`. Each user in the batch calls the `add` function registering to the coordinator and triggering the `Coordination Plan`.
+If more then one of the `Coordination Items` matching the `NumberOfUsers` is triggered, each of them will be passed a respective number of users.
+For example if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on the 6th user calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.
+
+`Coordination Actions` may be one of the following:
+ - `fun(Event) -> any()` - this type of action does not care about particular users, but only about the number of them;
+ - `fun(Event, ListOfUsersData) -> any()` - this type of action gets `ListOfUsersData` which is a list of tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+- `fun(Event, User1, User2) -> any()` - this type of action gets `distinct pairs` from the batch of users `User1` and `User2` which are tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+
+`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
+
+`distinct pairs` - in the context, these are pairs from collection where:
+ - when `{A, B}` is in the `distinct pairs` then `{B, A}` is not;
+ - when `{A, A}` is not in the `distinct pairs`;
+ - all pairs are distinct;
+ - Eg. for `[a]` `distinct pairs` collection is `[{a, undefined}]`;
+ - Eg. for `[a, b]` `distinct pairs` collection is `[{a, b}]`;
+ - Eg. for `[a, b, c]` `distinct pairs` collection is `[{a, b}, {a, c}, {b, c}]`.
+
+## Exports
+
+#### `start(CoordinatorName, CoordinationPlan) -> ok | error`
+#### `start(CoordinatorName, CoordinationPlan, Timeout) -> ok | error`
+
+##### Types
+```erlang
+CoordinatorName :: atom()
+CoordinationPlan  :: [ CoordinationItem ]
+CoordinationItem :: {NumberOfUsers :: pos_integer() | all, CoordinationActions}
+CoordinationActions :: [ CoordinationActions ] | CoordinationAction
+CoordinationAction :: 
+    fun(Event) -> any() |
+    fun(Event, ListOfUsersData) -> any() |
+    fun(Event, MaybeUser, MaybeUser) -> any()
```

Comment:
I understand that the two users given to this function are ensured to be different, right?
Then maybe you can write
```erlang
    fun(Event, MaybeUser1, MaybeUser2) when MaybeUser1 =/= MaybeUser2 -> any()
```

Reply:
@NelsonVides this is a type definition, so the users (the instances of data) might actually be different, but the type is the same nonetheless.

Reply:
```
$ erl -pa _build/default/lib/amoc/ebin/
Erlang/OTP 21 [erts-10.3.5.6] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1]

Enabled docsh 0.7.1 from: /Users/erszcz/work/erszcz/docsh/_build/default/lib/docsh
Call h(docsh) for interactive help.

Eshell V10.3.5.6  (abort with ^G)
1> l(amoc_coordinator).
{module,amoc_coordinator}
2> h(amoc_coordinator).
Module description is not available.
```

:'(

```
ok
3> t(amoc_coordinator).
-type coordination_action() :: fun((coordination_event(),
                                    [coordination_data()]) -> any()) |
                               fun((coordination_event(),
                                    maybe_coordination_data(),
                                    maybe_coordination_data()) -> any()) |
                               fun((coordination_event()) -> any()).
-type coordination_actions() :: [coordination_action()] | coordination_action().
-type coordination_data() :: {pid(), Data :: any()}.
-type coordination_event() :: coordinate | timeout | stop | reset.
-type coordination_item() :: {NoOfUsers :: pos_integer() | all,
                              coordination_actions()}.
-type coordination_plan() :: [coordination_item()] | coordination_item().
-type coordination_timeout_in_sec() :: pos_integer() | infinity.
-type maybe_coordination_data() :: coordination_data() | undefined.
-type normalized_coordination_item() :: {NoOfUsers :: pos_integer() | all,
                                         [coordination_action()]}.
-type state() :: {worker, pid()} | {timeout, pid()}.

ok
4>
```

:+1: